lib/index.js

```diff
@@ -1,21 +1,65 @@
-export default (callback) => (input) => {
+export default (opts = {}, callback) => (input) => {
   return function watch(log) {
     const chokidar = require('chokidar');
 
+    /* eslint-disable no-param-reassign, immutable/no-mutation */
+
+    if (!callback && typeof opts === 'function') {
+      callback = opts;
+      opts = {};
+    }
+
+    if (typeof opts === 'string' || Array.isArray(opts)) {
+      opts = { files: opts };
+    }
+
+    if (input && input.length && !opts.files) {
+      opts.files = input;
+    }
+
+    opts.events = opts.events || [ 'add', 'change' ];
+
+    /* eslint-enable */
+
     return new Promise((resolve, reject) => {
-      const init = () => {
-        chokidar
-          .watch(input, { persistent: true })
-          .on('change', callback)
-          .on('error', reject)
-          .on('ready', () => {
-            log('press ctrl-c to exit');
-          });
-      };
-
-      callback(input)
-        .then(init)
-        .catch(init);
+
+      log('Globbing files...');
+
+      const initialFileList = [];
+      const initialListener = (file) => initialFileList.push(file);
+
+      const watcher = chokidar.watch(opts.files, { persistent: true, ...opts });
+
+      watcher.once('error', reject);
+
+      watcher.on('add', initialListener);
+      watcher.once('ready', () => {
+        watcher.removeListener('add', initialListener);
+
+        log('initial scan complete, running task...');
+
+        const onDone = (err) => {
+          if (err && opts.bail) {
+            reject(err);
+          } else {
+            log('watching for changes, press ctrl-c to exit');
+            opts.events.forEach((event) => watcher.on(event, callback));
+            resolve();
+          }
+        };
+
+        try {
+          const ret = callback(initialFileList);
+
+          if (ret && ret.then) {
+            ret.then(onDone).catch(onDone);
+          } else {
+            onDone();
+          }
+        } catch (err) {
+          onDone(err);
+        }
+      });
```

Comment:
why not just to use a returned promise as is? https://github.com/start-runner/watch/blob/ea7ccd1231f1b1a081be737d749697fe2cda840d/lib/index.js#L16-L18